File: 2021/Day 5/Javascript/code.js
// Including this because it's pretty :)
function visualize(map, toString) {
    map = Array.from(map);
    const maxLength = Math.max(...map.map(a => (a || []).length || 0).filter(a => a >= 0));
    for (y = 0; y < map.length; y++) {
        for (x = 0; x < maxLength; x++) {
            map[y] = map[y] || [];
            map[y][x] = map[y][x] || 0;
        }
        if (toString) map[y] = map[y].join(", ");
    }
    if (toString) console.log(map.join("\n"));
    else console.table(map);
}

exports.part1 = (input) => {
    const point = input.split('\n');
    const map = [];
    for (let raw of point) {
        let [origin, destination] = raw.split(" -> ").map(x => x.split(",").map(Number));

        function addPoint(x, y) {
            map[y] = map[y] || [];
            map[y][x] = map[y][x] || 0;
            map[y][x]++;
        }

        if (origin[0] != destination[0] && origin[1] != destination[1]) continue;

        if (origin[0] == destination[0]) {
            for (let i = origin[1]; i <= destination[1]; i++) addPoint(origin[0], i);

            for (let i = origin[1]; i >= destination[1]; i--) addPoint(origin[0], i);
        }

        if (origin[1] == destination[1]) {
            for (let i = origin[0]; i <= destination[0]; i++) addPoint(i, origin[1]);

            for (let i = origin[0]; i >= destination[0]; i--) addPoint(i, origin[1]);
        }
    }
    return map.map(m => m.filter(x => x > 1)).reduce((p, v) => p += v.filter(x => (Number(x) != null) && x > 1).length, 0);
}

exports.part2 = (input) => {
    const point = input.split('\n');
    const map = [];
    for (let raw of point) {
        let [origin, destination] = raw.split(" -> ").map(x => x.split(",").map(Number));

        function addPoint(x, y) {
            map[y] = map[y] || [];
            map[y][x] = map[y][x] || 0;
            map[y][x]++;
        }

        if (origin[0] != destination[0] && origin[1] != destination[1]) {
            for (let x = origin[0]; x <= destination[0]; x++) addPoint(x, origin[1] > destination[1] ? origin[1] - Math.abs(x - origin[0]) : origin[1] + Math.abs(x - origin[0]));

            for (let x = origin[0]; x >= destination[0]; x--) addPoint(x, origin[1] > destination[1] ? origin[1] - Math.abs(x - origin[0]) : origin[1] + Math.abs(x - origin[0]));
        } else {
            if (origin[1] == destination[1]) {
                for (let i = origin[0]; i <= destination[0]; i++) addPoint(i, origin[1]);

                for (let i = origin[0]; i >= destination[0]; i--) addPoint(i, origin[1]);
            }

            if (origin[0] == destination[0]) {
                for (let i = origin[1]; i <= destination[1]; i++) addPoint(origin[0], i);

                for (let i = origin[1]; i >= destination[1]; i--) addPoint(origin[0], i);
            }
        }
    }
    return map.map(m => m.filter(x => x > 1)).reduce((p, v) => p += v.filter(x => (Number(x) != null) && x > 1).length, 0);
}

File: 2021/Day 5/lua/code.lua
function dofile(filename)
	return assert(loadfile(filename))()
end

local Array = dofile("../../Common\\lua\\Array.lua")
local Utility = dofile("../../Common\\lua\\Utility.lua")

local code = {}

function code:part1(input)
	local point = Utility.split(input, "\n", true)
	local map = Array()
	for _, raw in pairs(point) do
		local res = Utility.split(raw, " -> ", true):map(function(x)
			return Utility.split(x, ",", true):map(function(x)
				return tonumber(x)
			end)
		end)

		local origin = res[1]
		local destination = res[2]

		function addPoint(x, y)
			x = x + 1
			y = y + 1
			map[y] = map[y] or Array()
			map[y][x] = map[y][x] or 0
			map[y][x] = map[y][x] + 1
		end

		if origin[1] == destination[1] or origin[2] == destination[2] then
			if origin[1] == destination[1] then
				do
					local i = origin[2]
					while i <= destination[2] do
						addPoint(origin[1], i)
						i = i + 1
					end
				end

				do
					local i = origin[2]
					while i >= destination[2] do
						addPoint(origin[1], i)
						i = i - 1
					end
				end
			end

			if origin[2] == destination[2] then
				do
					local i = origin[1]
					while i <= destination[1] do
						addPoint(i, origin[2])
						i = i + 1
					end
				end
				do
					local i = origin[1]
					while i >= destination[1] do
						addPoint(i, origin[2])
						i = i - 1
					end
				end
			end
		end
	end
	return map
		:map(function(m)
			return m:filter(function(x)
				return x > 1
			end)
		end)
		:reduce(function(p, v)
			return p + v:filter(function(x)
				return tonumber(x) ~= nil and x > 1
			end).length
		end, 0)
end

function code:part2(input)
	local point = Utility.split(input, "\n", true)
	local map = Array()
	for _, raw in pairs(point) do
		local res = Utility.split(raw, " -> ", true):map(function(x)
			return Utility.split(x, ",", true):map(function(x)
				return tonumber(x)
			end)
		end)

		local origin = res[1]
		local destination = res[2]

		function addPoint(x, y)
			x = x + 1
			y = y + 1
			map[y] = map[y] or Array()
			map[y][x] = map[y][x] or 0
			map[y][x] = map[y][x] + 1
		end

		if origin[1] == destination[1] or origin[2] == destination[2] then
			if origin[1] == destination[1] then
				do
					local i = origin[2]
					while i <= destination[2] do
						addPoint(origin[1], i)
						i = i + 1
					end
				end

				do
					local i = origin[2]
					while i >= destination[2] do
						addPoint(origin[1], i)
						i = i - 1
					end
				end
			end

			if origin[2] == destination[2] then
				do
					local i = origin[1]
					while i <= destination[1] do
						addPoint(i, origin[2])
						i = i + 1
					end
				end
				do
					local i = origin[1]
					while i >= destination[1] do
						addPoint(i, origin[2])
						i = i - 1
					end
				end
			end
		else
			do
				local x = origin[1]
				while x <= destination[1] do
					addPoint(
						x,
						origin[2] > destination[2] and (origin[2] - math.abs(x - origin[1]))
							or (origin[2] + math.abs(x - origin[1]))
					)
					x = x + 1
				end
			end
			do
				local x = origin[1]
				while x >= destination[1] do
					addPoint(
						x,
						origin[2] > destination[2] and (origin[2] - math.abs(x - origin[1]))
							or (origin[2] + math.abs(x - origin[1]))
					)
					x = x - 1
				end
			end
		end
	end
	return map
		:map(function(m)
			return m:filter(function(x)
				return x > 1
			end)
		end)
		:reduce(function(p, v)
			return p + v:filter(function(x)
				return tonumber(x) ~= nil and x > 1
			end).length
		end, 0)
end

return code
